packages/ckeditor5-html-support/src/integrations/table.js

```diff
@@ -0,0 +1,164 @@
+/**
+ * @license Copyright (c) 2003-2021, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module html-support/integrations/table
+ */
+
+import { Plugin } from 'ckeditor5/src/core';
+import { disallowedAttributesConverter } from '../converters';
+import { setViewAttributes } from '../conversionutils.js';
+
+import DataFilter from '../datafilter';
+
+/**
+ * Provides the General HTML Support integration with {@link module:table/table~Table Table} feature.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class TableElementSupport extends Plugin {
+	static get requires() {
+		return [ DataFilter ];
+	}
+
+	init() {
+		const editor = this.editor;
+
+		if ( !editor.plugins.has( 'TableEditing' ) ) {
+			return;
+		}
+
+		const schema = editor.model.schema;
+		const conversion = editor.conversion;
+		const dataFilter = editor.plugins.get( DataFilter );
+
+		dataFilter.on( 'register:table', ( evt, definition ) => {
+			if ( definition.model !== 'table' ) {
+				return;
+			}
+
+			schema.extend( 'table', {
+				allowAttributes: [
+					'htmlAttributes',
+					// Figure, thead and tbody elements don't have model counterparts.
+					// We will be preserving attributes on table element using these attribute keys.
+					'htmlFigureAttributes', 'htmlTheadAttributes', 'htmlTbodyAttributes'
+				]
+			} );
+
+			conversion.for( 'upcast' ).add( disallowedAttributesConverter( definition, dataFilter ) );
+			conversion.for( 'upcast' ).add( viewToModelTableAttributeConverter( dataFilter ) );
+			conversion.for( 'downcast' ).add( modelToViewTableAttributeConverter() );
+
+			evt.stop();
+		} );
+
+		dataFilter.on( 'register:figure', () => {
+			conversion.for( 'upcast' ).add( consumeTableFigureConverter() );
+		} );
+	}
+}
+
+// View-to-model conversion helper preserving allowed attributes on {@link module:table/table~Table Table}
+// feature model element.
+//
+// @private
+// @param {module:html-support/datafilter~DataFilter} dataFilter
+// @returns {Function} Returns a conversion callback.
+function viewToModelTableAttributeConverter( dataFilter ) {
+	return dispatcher => {
+		dispatcher.on( 'element:table', ( evt, data, conversionApi ) => {
+			const viewTableElement = data.viewItem;
+
+			preserveElementAttributes( viewTableElement, 'htmlAttributes' );
+
+			const viewFigureElement = viewTableElement.parent;
+			if ( viewFigureElement.is( 'element', 'figure' ) ) {
+				preserveElementAttributes( viewFigureElement, 'htmlFigureAttributes' );
+			}
+
+			for ( const childNode of viewTableElement.getChildren() ) {
+				if ( childNode.is( 'element', 'thead' ) ) {
+					preserveElementAttributes( childNode, 'htmlTheadAttributes' );
+				}
+
+				if ( childNode.is( 'element', 'tbody' ) ) {
+					preserveElementAttributes( childNode, 'htmlTbodyAttributes' );
+				}
+			}
+
+			function preserveElementAttributes( viewElement, attributeName ) {
```

Comment:
Maybe we can move this function directly to the body of `viewToModelTableAttributeConverter()`? Otherwise it'll be created every time the converter for `element:table` is called.

Reply:
Extracting these functions outside the scope of the dispatcher function will require passing multiple (~5) parameters which looks ugly. We are looking into micro pre-optimization here, so I would prefer to keep it more readable instead.